### PR TITLE
work without ssh, option for port

### PR DESCRIPTION
**Pull request by deitch**
This started out as just a proof of concept that these things are doable, but became a working element of some of it.

There is more to be done, but this does:

* option for port (default to 9000)
* option to use private IP only
* sets everything up via user-data rather than ssh
* waits for the health-check to be available
* makes the user-data `set -e` so it exits on immediate error, and any fixes necessary to make that work
* minor: reorders `vars.template` to match the order of `variables.tf`, just because it is easier to read

### CODE REVIEW

File: assets/user_data.sh [hunk 1]
```diff
@@ -1,5 +1,8 @@
 #!/bin/bash
```
Comment (displague) [inline assets/user_data.sh:1]: Do you find it confusing to think of this as a `.sh` file when it is a template file? Perhaps we should rename this to `.sh.tpl` and/or lead-in with a comment about how this file will be interpreted by Terraform as a template (listing all of the expected variables).

Comment (deitch) [reply assets/user_data.sh:1]: Yeah, that is a pretty good idea. It predates this PR, but might as well get it in. I think it makes sense to leave it as `.sh` to make it easier for IDEs/emacs/vi, though. I can add a comment.

Comment (deitch) [reply assets/user_data.sh:1]: Done. Take a look.

File: assets/user_data.sh [hunk 2]
```diff
@@ -52,6 +55,15 @@ dist-check
   # Install prerequisites
   install-prerequisites
 
+# figure out our bind port
+IS_PUBLIC=${public}
+# by default, listen to all IP
+LISTEN_IP=
+# if not public, grab the private IP only
+if [ -z "$IS_PUBLIC" ]; then
```
[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh:67]: Sure. That can be updated.

[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh:67]: I don't love the idea of `TF_TPL` as a prefix. But I do like the idea of pulling everything that needs to be interpolated into one part of the file and setting env vars, to make it easier to read.

[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh]: Addressed this particular issue (`IS_PUBLIC` -> `PUBLIC`).

[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh]: And moved all of the interpolations to the beginning

**Comment by displague** [comment]
This is great, @deitch. I like the move away from SSH provisioners.

I've pushed a commit that updates this project to TF 0.13 (so that the latest and any future Packet provider can be consumed).

I ran into a problem testing this (which we discussed offline, but I'll repeat here).  The health checks timed out. This appears to be due to the hostnames no longer being defined (in /etc/hosts).

We discussed a few alternatives:
* use {unhyphenated-deviceid}.packethost.net names (this can not be known at userdata provision time)
* use IP addresses (this can not be known at userdata provision time)
* bringing back the /etc/hosts null resource ssh provisioner (let's not)
* create a private IP pool and assign well known (independent) addresses based on count.index

The last approach (your idea 🧠 ) is the one that sounds best.

---

I'm leary of the (existing) dependency on a local `curl` binary for the health checks, this would not work in Windows environments and perhaps some Linux environments.  I went looking for alternatives, like a file provisioner that uses `http` or a generic "health check" provider (perhaps https://registry.terraform.io/providers/hashicorp/http/latest/docs/data-sources/data_source?). Nothing came up.  This is fodder for a future issue, not a concern now.

**Comment by deitch** [comment]
> I'm leary of the (existing) dependency on a local curl binary for the health checks, this would not work in Windows environments and perhaps some Linux environments

Agreed. I couldn't think of anything better offhand, though. As you said, let's leave that part for future discussion. I do wish terraform had a "health-check" option.

**Comment by deitch** [comment]
> create a private IP pool and assign well known (independent) addresses based on count.index

I just realized something. Packet elastic IP pools only work for public IPv4 or global IPv4; there is no private IPv4. 🤦🏻‍♂️

**Comment by deitch** [comment]
And I realized we have an issue with the first two. Beyond the problem that we do not know the hostnames/IPs at userdata provision time, if those changed (e.g. we add or remove a host), the originally installed ones have no way to access them. E.g. one of the 4 minio in a cluster dies, so we provision a new one. The new one has a new IP address.  But the existing four cannot reach `myminio-4` because their `/etc/hosts` is out of date.

I am a bit stumped.

**Comment by deitch** [comment]
The right way to do this is just to have DNS entries. You know the name in advance, and you are done. But Packet has no DNS.

**Comment by deitch** [comment]
I ended up using public EIPs, which we can create before and then pass on to userdata, which can be passed to instance creation. No ssh needed. Less brittle. I prefer private EIPs or dynamic DNS, but this is good enough to move ahead.

cc @displague

File: assets/user_data.tpl.sh [hunk 3]
```diff
@@ -167,77 +133,76 @@ set_partition() {
...
-    mount "$${MOUNTPOINT}"
-done
-
-# Amount of disks that got partitioned
-echo "$${#DISKS[@]}"
+# Install pre-requisites (parted and filesystem packages)
+install-prerequisites() {
+  # Installation begins here
+  # shellcheck disable=SC2235
+  if [ "$DISTRO" == "ubuntu" ]; then
```
Comment (displague) [inline assets/user_data.tpl.sh:140]: This could be trimmed down with a `case/esac` statement.

Comment (deitch) [reply assets/user_data.tpl.sh:140]: It could indeed. I didn't write it, but happy to clean it if we already are PRing (is that a verb?)

Comment (deitch) [reply assets/user_data.tpl.sh]: Done

**Comment by deitch** [comment]
I also managed to separate out the templating, so that the userdata script is _much_ cleaner. Have a look @displague

File: assets/user_data.sh [hunk 4]
```diff
@@ -12,232 +27,177 @@ function dist-check() {
...
-done
-
-# Amount of disks that got partitioned
-echo "$${#DISKS[@]}"
+# Install pre-requisites (parted and filesystem packages)
+install-prerequisites() {
+  # Installation begins here
+  # shellcheck disable=SC2235
+  case "$DISTRO" in
+    ubuntu)
```
[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh:132]: You want _real_ simplicity?

[the remark this answers is not in the document]
Comment (deitch) [reply assets/user_data.sh]: Done. Look now @displague

File: README.md [hunk 5]
```diff
@@ -11,17 +11,17 @@ If you require support, please email [support@packet.com](support@packet.com), v
...
- 
-Here is an example for **macOS**: 
-```bash 
-curl -LO https://releases.hashicorp.com/terraform/0.12.25/terraform_0.12.25_darwin_amd64.zip 
+## Install Terraform
+Terraform is just a single binary.  Visit their [download page](https://www.terraform.io/downloads.html), choose your operating system, make the binary executable, and move it into your path.
+
+Here is an example for **macOS**:
+```bash
+curl -LO https://releases.hashicorp.com/terraform/0.12.25/terraform_0.12.25_darwin_amd64.zip
```
Comment (displague) [inline README.md:19]: I should have updated this in my 0.13 commit. What do you think @deitch - update it or defer the "how to install terraform" instructions to a Terraform URL?

File: variables.tf [hunk 6]
```diff
@@ -42,6 +36,16 @@ variable "hostname" {
     default = "minio-storage-node"
 }
 
+variable "port" {
+  description = "Port minio will listen on"
+  default = 9000
+}
+
+variable "public" {
+  description = "Listen on public IPv4"
```
Comment (displague) [inline variables.tf:45]: I like how this is a simple toggle that we can use if we introduce L2 private IP space later.

Comment (displague) [reply variables.tf:45]: Even if it doesn't do anything now: https://github.com/packet-labs/terraform-packet-distributed-minio/pull/13/files#diff-d005ed3614daa0c76778a7ed17279aa0R280-R283

File: README.md (hunk 5 above)
Comment (deitch) [reply README.md:19]: Yeah I will update that.

Comment (deitch) [reply README.md]: Just updated.

**Pull request merged**